src/compiler/scala/tools/nsc/transform/Erasure.scala

```diff
@@ -1155,6 +1196,17 @@ abstract class Erasure extends AddInterfaces
             case ArrayValue(elemtpt, trees) =>
               treeCopy.ArrayValue(
                 tree1, elemtpt setType specialScalaErasure.applyInArray(elemtpt.tpe), trees map transform).clearType()
+            case ValDef(_, _, tpt, rhs) =>
+              try super.transform(tree1).clearType()
+              finally {
+                tpt.tpe match {
+                  case FoldableConstantType(_) if !rhs.isInstanceOf[Literal] =>
+                    val deconst = tpt.tpe.deconst
+                    tpt setType deconst
+                    tree1.symbol.setInfo(deconst)
```

Comment:
This section of code assumes that `super.transform(tree1)` resulted in a tree with `tpt` was preserved. That's likely to be the case looking at the other cases in the transform, but it seems somewhat fragile. I'd prefer:

```
val vd1 = super.transform(....).clearType().asInstanceOf[ValDef]
vd1.tpt1 match { ... }
vd1
```